File: 2658-maximum-number-of-fish-in-a-grid/2658-maximum-number-of-fish-in-a-grid.cs
public class Solution {
    (int x, int y)[] dirs = new []{(0, -1), (0, 1), (-1, 0), (1, 0)};

    public int FindMaxFish(int[][] grid) {
        int m = grid.Length, n = grid[0].Length;
        var visited = new bool[m,n];
        var maxFish = 0;

        for(var i=0; i<m; i++){
            for(var j=0; j<n; j++){
                if(grid[i][j] == 0 || visited[i,j]) continue;
                maxFish = Math.Max(maxFish, BFS(grid, visited, i, j));
            }
        }

        return maxFish;
    }

    private int BFS(int[][] grid, bool[,] visited, int x, int y){
        int m = grid.Length, n = grid[0].Length;

        var queue = new Queue<(int, int)>();
        queue.Enqueue((x, y));
        visited[x, y] = true;
        var count = grid[x][y];

        while(queue.Count > 0){
            (x, y) = queue.Dequeue();
            foreach(var (dx, dy) in dirs){
                int xx = x + dx, yy = y + dy;
                if(xx == -1 || yy == -1 || xx == m || yy == n || grid[xx][yy] == 0 || visited[xx,yy]) continue;
                queue.Enqueue((xx, yy));
                visited[xx, yy] = true;
                count += grid[xx][yy];
            }
        }

        return count;
    }
}

File: 2658-maximum-number-of-fish-in-a-grid/2658-maximum-number-of-fish-in-a-grid.java
class Solution {
    int[][] dirs = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};

    public int findMaxFish(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        var visited = new boolean[m][n];
        var maxFish = 0;

        for(var i=0; i<m; i++){
            for(var j=0; j<n; j++){
                if(grid[i][j] == 0 || visited[i][j]) continue;
                maxFish = Math.max(maxFish, bfs(grid, visited, i, j));
            }
        }

        return maxFish;
    }

    private int bfs(int[][] grid, boolean[][] visited, int x, int y){
        int m = grid.length, n = grid[0].length;

        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{x, y});
        visited[x][y] = true;
        var count = grid[x][y];

        while(!queue.isEmpty()){
            var cords = queue.poll();
            x = cords[0]; y = cords[1];
            for(var dir : dirs){
                int dx = dir[0], dy = dir[1];
                int xx = x + dx, yy = y + dy;
                if(xx == -1 || yy == -1 || xx == m || yy == n || grid[xx][yy] == 0 || visited[xx][yy]) continue;
                queue.offer(new int[]{xx, yy});
                visited[xx][yy] = true;
                count += grid[xx][yy];
            }
        }

        return count;
    }
}